### PR TITLE
channeldb+rpc: add pagination of listinvoices #1514

### PR DESCRIPTION
**Pull request by taketa**
Closes: #1514
Short description: once we exceed a few thousand invoices, we can no longer return them over the RPC interface in a single go. As a result, we need to add pagination in order to allow callers to iteratively request the entire range by using time slice.

### CODE REVIEW

File: lnrpc/rpc.proto [hunk 1]
```diff
@@ -1708,9 +1708,28 @@ message PaymentHash {
...
+
+    /// EndTime is the end time of the time slice.
+    int64 end_time = 4;
+
+    /// IndexOffset is the offset within the time slice to start at. This
+    /// can be used to start the response at a particular record.
+    uint32 index_offset = 5;
+
+    /// NumMaxEvents is the max number of events to return.
+    uint32 num_mux_events = 6;
```
Comment (wpaulino) [inline lnrpc/rpc.proto]: s/mux/max/

File: cmd/lncli/commands.go [hunk 2]
```diff
@@ -2268,6 +2268,23 @@ var listInvoicesCommand = cli.Command{
 			Usage: "toggles if all invoices should be returned, or only " +
 				"those that are currently unsettled",
 		},
+		cli.IntFlag{
+			Name: "start",
+			Usage: "is the start time (" +
```
Comment (wpaulino) [inline cmd/lncli/commands.go]: Expand line to 80 chars. The same applies below.

File: cmd/lncli/commands.go [hunk 3]
```diff
@@ -2283,6 +2300,12 @@ func listInvoices(ctx *cli.Context) error {
 
 	req := &lnrpc.ListInvoiceRequest{
 		PendingOnly: pendingOnly,
+		StartTime: int64(ctx.Int("start")),
+		EndTime: int64(ctx.Int("end")),
+		IndexOffset: uint32(ctx.Int("offset")),
+		NumMuxEvents:50,
```
Comment (wpaulino) [inline cmd/lncli/commands.go]: This should also be configurable by the user.

File: rpcserver.go [hunk 4]
```diff
@@ -2825,30 +2825,59 @@ func (r *rpcServer) LookupInvoice(ctx context.Context,
...
+		
+		for _, dbInvoice := range dbInvoices{
+			creationDate:=dbInvoice.CreationDate.Unix()
+			
+			if creationDate>=req.StartTime && creationDate<=req.EndTime{
+				invoicesTimeSlice = append(invoicesTimeSlice,dbInvoice)
+			}
+		}
+		
+		if invoicesTimeSlice==nil{
```
Comment (wpaulino) [inline rpcserver.go]: What's the rationale behind this? If we didn't find any invoices within the time series, we shouldn't return any.

File: rpcserver.go [hunk 5]
```diff
@@ -2825,30 +2825,59 @@ func (r *rpcServer) LookupInvoice(ctx context.Context,
...
+			invoicesTimeSlice []channeldb.Invoice
+			invoices []*lnrpc.Invoice
+		)
+	
+		dbInvoices, err := r.server.chanDB.FetchAllInvoices(req.PendingOnly)
 		if err != nil {
 			return nil, err
 		}
+		
+		for _, dbInvoice := range dbInvoices{
```
Comment (wpaulino) [inline rpcserver.go]: Most of the filtering logic should be moved out of the `rpcserver` for easier testability. As pointed out in the original issue, we want to replicate the design of the [forwarding events](https://github.com/lightningnetwork/lnd/blob/master/channeldb/forwarding_log.go#L156) within the switch.

**Comment by wpaulino** [comment]
Also, you should reach out to Travis to unban your account so that the tests can run on the next push.

**Comment by Roasbeef** [comment]
Ping @wpaulino.

**Comment by wpaulino** [comment]
Replaced by #1720.

**Pull request closed without merging**